### PR TITLE
Add support for messages with type purchase_notification

### PR DESCRIPTION
**Pull request by Soheab**
## Summary

Relevant docs PR/Discussions:
- https://github.com/discord/discord-api-docs/pull/6927
- https://github.com/discord/discord-api-docs/discussions/6935#discussioncomment-9786723

## Checklist

- [ ] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/message.py [hunk 1]
```diff
@@ -837,6 +841,43 @@ def __init__(self, data: RoleSubscriptionDataPayload) -> None:
...
+    guild_product_purchase: Optional[:class:`GuildProductPurchase`]
+        The guild product purchase that prompted the message.
+    """
+
+    def __init__(self, data: PurchaseNotificationResponsePayload) -> None:
+        self._type: int = data['type']
+
+        self.guild_product_purchase: Optional[GuildProductPurchase] = None
+        guild_product_purchase = data.get('guild_product_purchase')
+        if guild_product_purchase is not None:
```
[the remark this answers is not in the document]
Comment (Soheab) [reply discord/message.py:874]: The value can be null according to the [unofficial api docs](https://docs.discord.sex/resources/message#message-purchase-notification-structure) so I guess I don't need to use .get()

File: discord/message.py [hunk 2]
```diff
@@ -2415,6 +2469,11 @@ def system_content(self) -> str:
         if self.type is MessageType.guild_incident_report_false_alarm:
             return f'{self.author.name} reported a false alarm in {self.guild}.'
 
+        if self.type is MessageType.purchase_notification and self.purchase_notification is not None:
+            guild_product_purchase = self.purchase_notification.guild_product_purchase
+            if guild_product_purchase is not None:
```
[the remark this answers is not in the document]
Comment (Soheab) [reply discord/message.py:2474]: But it can be None?

File: discord/message.py [hunk 3]
```diff
@@ -837,6 +841,45 @@ def __init__(self, data: RoleSubscriptionDataPayload) -> None:
...
+class GuildProductPurchase(NamedTuple):
+    """Represents a message's guild product that the user has purchased.
+
+    .. versionadded:: 2.5
+    """
+
+    listing_id: int
+    """The ID of the listing that the user has purchased."""
+    product_name: str
+    """The name of the product that the user has purchased."""
```
Comment (Rapptz) [inline discord/message.py:853]: We don't use NamedTuple in the public API if we can help it. Please just turn this into a regular class.

File: docs/api.rst [hunk 4]
```diff
@@ -5198,6 +5204,22 @@ RoleSubscriptionInfo
 .. autoclass:: RoleSubscriptionInfo
     :members:
 
+PurchaseNotification
+~~~~~~~~~~~~~~~~~~~~~
+
+.. attributetable:: PurchaseNotification
+
+.. autoclass:: PurchaseNotification
```
Comment (Rapptz) [inline docs/api.rst:5212]:
```suggestion
.. autoclass:: PurchaseNotification()
```

File: docs/api.rst [hunk 5]
```diff
@@ -5198,6 +5204,22 @@ RoleSubscriptionInfo
...
+    :members:
+
+GuildProductPurchase
++++++++++++++++++++++
+
+.. attributetable:: GuildProductPurchase()
```
Comment (Rapptz) [inline docs/api.rst:5218]:
```suggestion
.. attributetable:: GuildProductPurchase
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/message.py [hunk 6]
```diff
@@ -843,6 +847,51 @@ def __init__(self, data: RoleSubscriptionDataPayload) -> None:
...
+    product_name: :class:`str`
+        The name of the product that the user has purchased.
+    """
+
+    def __init__(self, data: GuildProductPurchasePayload) -> None:
+        self.listing_id: int = int(data['listing_id'])
+        self.product_name: str = data['product_name']
+
+    def __hash__(self) -> int:
+        return self.listing_id >> 22
```
Comment (Rapptz) [inline discord/message.py:868]: `__hash__` with no `__eq__` is weird, you should define both if you're going to do that.

File: discord/message.py [hunk 7]
```diff
@@ -843,6 +847,51 @@ def __init__(self, data: RoleSubscriptionDataPayload) -> None:
...
+    .. versionadded:: 2.5
+
+    Attributes
+    -----------
+    listing_id: :class:`int`
+        The ID of the listing that the user has purchased.
+    product_name: :class:`str`
+        The name of the product that the user has purchased.
+    """
+
```
Comment (Rapptz) [inline discord/message.py:862]: This is missing `__slots__`.

**Review by Rapptz — CHANGES_REQUESTED** [review]
Just some minor nits

**Review by Rapptz — APPROVED** [review]

**Pull request merged**